webpack.config.ts

```diff
@@ -0,0 +1,189 @@
+import path from 'path';
+import webpack from 'webpack';
+import ThreadsPlugin from 'threads-plugin';
+import nodeExternals from 'webpack-node-externals';
+import DeclarationBundlePlugin from './webpack-plugins/DeclarationBundlePlugin';
+
+function resolveProto(context, request: string, callback) {
+  if (/.*\/proto\/js\/.*/.test(request)){
+    // Extract name of js file
+    const fileName = request.split('/').pop()
+    // Externalize to a commonjs module using the request path
+    const filePath = path.resolve(process.cwd(), 'proto', 'js', fileName!)
+    return callback(null, 'commonjs ' + filePath);
+  }
+  callback();
+}
+
+const libraryConfig = {
+  name: 'polykey',
+  entry: './src/lib/Polykey.ts',
+  output: {
+    path: path.resolve(__dirname, 'dist', 'lib'),
+    filename: 'polykey.js',
+    library: 'polykey',
+    libraryTarget: 'umd',
+    globalObject: 'this'
+  },
+  devtool: "source-map",
+  resolve: {
+    extensions: ['.ts', '.js'],
+  },
+  target: 'node',
+  node: {
+    __dirname: false,
+    __filename: false
+  },
+  module: {
+    rules: [
+      {
+        test: /.tsx?$/,
+        loader: 'ts-loader'
+      }
+    ]
+  },
+  externals: [
+    resolveProto,
+    { fs: "commonjs fs" },
+    nodeExternals()
+  ],
+  plugins: [
+    new ThreadsPlugin({
+      globalObject: false
+    }),
+    new webpack.DefinePlugin({
+      PROJECT_DIR: path.resolve(__dirname)
+    }),
+    new DeclarationBundlePlugin({
+      entry: './src/lib/Polykey.ts',
+      output: './dist/lib/polykey.d.ts'
+    })
+  ],
+  watchOptions: {
+    ignored: /node_modules/
+  }
+}
+
+const cliConfig = {
+  name: 'cli',
+  entry: './src/cli/index.ts',
+  output: {
+    path: path.resolve(__dirname, 'dist', 'cli'),
+    filename: 'cli.js',
+    library: 'cli',
+    libraryTarget: 'umd',
+    globalObject: 'this'
+  },
+  resolve: {
+    extensions: ['.ts'],
+  },
+  target: "node",
+  node: {
+    __dirname: false,
+    __filename: false
+  },
+  module: {
+    rules: [
+      {
+        test: /.tsx?$/,
+        loader: 'ts-loader'
+      }
+    ]
+  },
+  externals: [
+    { "../agent/PolykeyAgent": "commonjs ../agent/agent.js" },
+    { "../lib/Polykey": "commonjs ../lib/Polykey.js" },
+    { fs: "commonjs fs" },
+    nodeExternals()
+  ],
+  plugins: [
+    new DeclarationBundlePlugin({
+      ignoreDeclarations: true
+    })
+  ],
+  watchOptions: {
+    ignored: /node_modules/
+  }
+}
+
+const agentConfig = {
```

Comment:
I'm wondering why there's so many separate builds here. I thought there would be:

1. a library build
2. a cli build

Especially since the cli can use subcommands to launch daemon or whatever.

Reply:
That is a good point, the agent code can be merged into the lib code. The daemon script is a little more challenging to separate because it is a static script to be called by `fork(..)`

Reply:
I managed to combine the agent and lib code, so now we are down to three separate builds. I could probably get it down to 2 if I just use webpacks copy plugin

Reply:
actually that probably wouldn't be easy since it has static imports that need to change with webpack externals.

Reply:
I might leave this as 3 unless we can come up with a way to compile scripts that will be `fork`ed with webpack.

Reply:
It is possible to "fork" with 1 cli script, by doing "self-forking". Basically the script calls itself. I've seen this pattern before, it's pretty ingenious. It's like recursion at the process level. But when you call yourself again, you provide parameters that understands that it is going to be a daemon.